pjrt/convert/trace_container.zig

```diff
@@ -0,0 +1,223 @@
+const std = @import("std");
+const xplane_proto = @import("//tsl:xplane_proto");
+const xplane_schema = @import("utils/xplane_schema.zig");
+const xplane_utils = @import("utils/xplane_utils.zig");
+const xplane_visitor = @import("utils/xplane_visitor.zig");
+
+// Constants used as trace_viewer PID (device_id in trace_events.proto).
+// PID 0 is unused.
+// Support up to 500 accelerator devices.
+const first_device_id = 1;
+const last_device_id = 500;
+// Support Upto 200 custom planes as fake devices (i.e., planes with a
+// "/custom:" prefix). See `<project_name>::custom_plane_prefix` for more
+// information
+const first_custom_plane_device_id = last_device_id + 1;
+const max_custom_plane_devices_per_host = 200;
+const last_custom_plane_device_id = first_custom_plane_device_id + max_custom_plane_devices_per_host - 1;
+
+// Host threads are shown as a single fake device.
+pub const host_threads_device_id = last_custom_plane_device_id + 1;
+
+pub const xla_async_op_line_name = "Async XLA Ops";
+
+pub const host_threads_plane_name = "/host:CPU";
+pub const gpu_plane_prefix = "/device:GPU:";
+pub const tpu_plane_prefix = "/device:TPU:";
+pub const custom_plane_prefix = "/device:CUSTOM:";
+
+pub const TraceContainer = struct {
+    metadata: Trace = .{},
+    events: std.ArrayListUnmanaged(*TraceEvent) = .{},
+
+    pub const Trace = struct {
+        devices: std.AutoHashMapUnmanaged(u32, Device) = .{},
+        trace_events: std.ArrayListUnmanaged(TraceEvent) = .{},
+    };
+
+    pub const TraceEvent = struct {
+        device_id: u32 = 0,
+        resource_id: u32 = 0,
+        name: []const u8 = &[_]u8{},
+        timestamp_ps: u128 = 0,
+        duration_ps: u64 = 0,
+        args: std.StringHashMapUnmanaged([]const u8) = .{},
+    };
+
+    pub const Device = struct {
+        name: []const u8 = &[_]u8{},
+        device_id: u32 = 0,
+        resources: std.AutoHashMapUnmanaged(u32, Resource) = .{},
+
+        pub fn mutableResource(self: *Device, allocator: std.mem.Allocator, resource_id: u32) !*Resource {
+            const entry = try self.resources.getOrPut(allocator, resource_id);
```

Comment:
I don't really like this pattern because it default inits value. I'd rather you use `getOrPutValue` directly from the caller.